README.md
Dockmaster
==========

Dockmaster who oversees the workings of containers throughout a network.

Dockmaster is a overseer for all the containers running on all the servers. It's service discovery for containers.

The architecture of Dockmaster is simple.

Server
------

Will sit on a central server which is reachable by all the nodes on which an Agent is operating.

Agent
-----

An Agent sits on a node where there are docker containers running. The Agent reports to the Server about statuses of the containers next-to it.


Currently the Agent only reports about running containers. Other containers are not regarded because it would create a very convoluted view.

The frontend first needs to handle proper showing of the data that it gets.

Install
=======

Install All dependencies

```bash
    go get -v ./...
```

Starting the server:

```bash
    cd server
    go build # To build the project
    ./server
```

Starting the agent:

```bash
    cd agent
    go build # To build the project
    ./agent # running it without flags will make it run with default values. For flags, please see the Agent's README.
```

Starting the frontend:

```bash
    cd frontend
    go build # To build the project
    ./frontend # The front-end is very basic. Defining a server is not optional. Currently it has to run where the server is running.
```

Current Look
------------

The current look is a very simple table view:

![Dockmaster](frontend1.png)

After clicking inspect, a popup windows will provide more information about the container.

![Popup](frontend2.png)

Bulk Stop
---------

Once you have an agent up and running, it is now possible to stop all running containers for that agent by clicking the 'Stop All' button next to the Agent.

![Stop All](frontend3.png)

The agent takes a minute to update its status, so the stopped container state, right now, will not be refreshed immediately. A message is received if the operation was successful or not. This, right now, is only displayed in the console output.

![Message](frontend4.png)

After the agent has refreshed its status, it can be seen that all the containers have been shutdown.

![No Containers Running For Agent](frontend5.png)

Stopping a single container is now easily achieved with the click of a button next to the Inspect link.

![Stopping a single Container](frontend6.png)


CHANGELOG.md
Changes to Dockmaster
=====================

v0.2.1
* Front-end adjusted to contain the new abilities of the Agent

v0.2.0
* Agent is now two-way. It can inspect containers and perform basic operations on it
* Inspect is available, others are under development


v0.1.1

* Started to change agent so that it has now two way communication
* Communication will be limited to simple actions for containers in the beginning

v0.1.0
------

 * Agent created
 * Server created
 * Front-end created


server/dockmaster_test.go
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestGetContainer(t *testing.T) {
	t.SkipNow()
}

func TestSaveContainer(t *testing.T) {
	t.SkipNow()
}

func TestGetConfiguration(t *testing.T) {
	c := getConfiguration()

	assert.NotEmpty(t, c)
}


server/mongodb.go
package main

import (
	"log"
	"time"

	"gopkg.in/mgo.v2"
	"gopkg.in/mgo.v2/bson"
)

//MongoDBConnection Encapsulates a connection to a database
type MongoDBConnection struct {
	session *mgo.Session
}

//Save will save a contaier using mongodb as a storage medium
func (mdb MongoDBConnection) Save(a Agent) error {
	mdb.session = mdb.GetSession()
	defer mdb.session.Close()
	db := mdb.session.DB("dockmaster").C("containers")
	db.Remove(bson.M{"agentid": a.AgentID})

	now := time.Now()
	//This is necessary to convert time.Now() to UTC which is CET by default
	date := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute(), now.Second(), now.Nanosecond(), time.UTC)
	a.CreatedAt = date
	err := db.Insert(a)
	if err != nil {
		return err
	}
	return nil
}

//removeOldData removes old data from the database on save
func (mdb MongoDBConnection) startCleansing() {
	log.Println("Looking for old Documents to Cleanse...")
	mdb.session = mdb.GetSession()
	defer mdb.session.Close()
	db := mdb.session.DB("dockmaster").C("containers")
	agents := []Agent{}
	iter := db.Find(nil).Iter()
	iter.All(&agents)
	now := time.Now()
	//time.Now stays in CET even after time.Now().UTC(). Which means this is needed to force it to UTC.
	date := time.Date(now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute(), now.Second(), now.Nanosecond(), time.UTC)
	for _, a := range agents {
		compareNow := a.CreatedAt.Add(time.Second * time.Duration(a.ExpireAfterSeconds)).UTC()
		if compareNow.Unix() < date.Unix() {
			log.Println("Cleansing Old Document:", a)
			db.Remove(bson.M{"agentid": a.AgentID})
		}
	}
}

//Load will load the contaier using mongodb as a storage medium
func (mdb MongoDBConnection) Load() (a []Agent, err error) {
	mdb.session = mdb.GetSession()
	defer mdb.session.Close()
	c := mdb.session.DB("dockmaster").C("containers")

	iter := c.Find(nil).Iter()
	err = iter.All(&a)

	return a, err
}

//GetAgent retrieves one agent to perform an operation with
func (mdb MongoDBConnection) GetAgent(agentID string) (a Agent, err error) {
	mdb.session = mdb.GetSession()
	defer mdb.session.Close()
	c := mdb.session.DB("dockmaster").C("containers")

	err = c.Find(bson.M{"agentid": agentID}).One(&a)
	return a, err
}

//Delete bulk deletes containers
func (mdb MongoDBConnection) Delete(a Agent) error {
	mdb.session = mdb.GetSession()
	defer mdb.session.Close()
	db := mdb.session.DB("dockmaster").C("containers")
	err := db.Remove(bson.M{"agentid": a.AgentID})
	if err != nil {
		return err
	}
	return nil
}

//GetSession return a new session if there is no previous one
func (mdb *MongoDBConnection) GetSession() *mgo.Session {
	if mdb.session != nil {
		return mdb.session.Copy()
	}
	session, err := mgo.Dial("localhost")
	if err != nil {
		panic(err)
	}
	session.SetMode(mgo.Monotonic, true)
	return session
}


frontend/frontend.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"html/template"
	"net/http"
	"path"
)

//Container a single container
type Container struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	RunningCmd string `json:"command"`
	Port       string `json:"port"`
}

//Agent post data from an agent with ID and containers it has.
type Agent struct {
	AgentID            string      `json:"agentid"`
	ExpireAfterSeconds int         `json:"expireAfterSeconds"`
	Containers         []Container `json:"containers"`
}

var server string

func init() {
	flag.StringVar(&server, "server", "http://localhost:8989", "Full URL for the server.")
	flag.Parse()
}

func index(w http.ResponseWriter, r *http.Request) {
	resp, _ := http.Get(server + "/api/1/list")
	agents := []Agent{}
	decoder := json.NewDecoder(resp.Body)
	err := decoder.Decode(&agents)
	if err != nil {
		fmt.Fprintf(w, "error occured:"+err.Error())
		return
	}

	lp := path.Join("templates", "layout.html")
	tmpl, err := template.ParseFiles(lp)
	if err != nil {
		fmt.Fprintf(w, "error occured:"+err.Error())
		return
	}
	tmpl.Execute(w, agents)
}

func main() {
	http.HandleFunc("/", index)
	fs := http.FileServer(http.Dir("static"))
	http.Handle("/static/", http.StripPrefix("/static/", fs))
	http.ListenAndServe(":9191", nil)
}


server/dockmaster.go
package main

import (
	"encoding/json"
	"net/http"
	"time"

	"github.com/fsouza/go-dockerclient"
	"github.com/gin-gonic/gin"
)

//APIVERSION Is the current API version
const APIVERSION = "1"

//APIBASE Defines the API base URI
const APIBASE = "api/" + APIVERSION

var mdb MongoDBConnection

//Container a single container
type Container struct {
	ID         string `json:"id"`
	Name       string `json:"name"`
	RunningCmd string `json:"command"`
	Port       string `json:"port"`
}

//Agent post data from an agent with ID and containers it has.
type Agent struct {
	AgentID            string      `json:"agentid"`
	ExpireAfterSeconds int         `json:"expireAfterSeconds"`
	Containers         []Container `json:"containers"`
	CreatedAt          time.Time   `bson:"createdAt"`
	IP                 string      `json:"ip"`
	Port               string      `json:"port"`
}

func init() {
	mdb = MongoDBConnection{}
	go func() {
		for {
			//This will remove old data periodically at every minute.
			//Mongodb's own cleanser also ticks at every 60 seconds.
			mdb.startCleansing()
			time.Sleep(time.Minute)
		}
	}()
}

func main() {
	router := gin.Default()
	v1 := router.Group(APIBASE)
	{
		v1.GET("/list", listContainers)
		v1.POST("/add", addContainers)
		v1.POST("/delete", deleteContainers)
		v1.GET("/inspect/:agentID/:containerID", inspectContainer)
		v1.OPTIONS("/inspect/:agentID/:containerID", preflight)
		v1.POST("/stopAll", stopAll)
		v1.OPTIONS("/stopAll", preflight)
		v1.POST("/stop/:agentID", stop)
		v1.OPTIONS("/stop/:agentID", preflight)
	}
	router.Run(":8989")
}

func preflight(c *gin.Context) {
	c.Header("Access-Control-Allow-Origin", "*")
	// c.Header("Access-Control-Allow-Methods", "GET, POST, OPTIONS")
	c.Header("Access-Control-Allow-Headers", "access-control-allow-origin, access-control-allow-headers")
	c.JSON(http.StatusOK, struct{}{})
}

func stopAll(c *gin.Context) {
	var agentID struct {
		AgentID string `json:"agentid"`
	}
	err := c.BindJSON(&agentID)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error binding json: " + err.Error()})
		return
	}

	a, err := mdb.GetAgent(agentID.AgentID)
	resp, _ := http.Post("http://"+a.IP+":"+a.Port+"/"+APIBASE+"/stopAll", "application/json", nil)

	c.Header("Access-Control-Allow-Origin", "*")
	c.Header("Access-Control-Allow-Headers", "access-control-allow-origin, access-control-allow-headers")
	if resp.StatusCode != 200 {
		c.JSON(resp.StatusCode, resp.Body)
		return
	}

	c.JSON(resp.StatusCode, Message{"All containers stopped."})
}

func stop(c *gin.Context) {
	agentID := c.Param("agentID")

	a, _ := mdb.GetAgent(agentID)
	resp, _ := http.Post("http://"+a.IP+":"+a.Port+"/"+APIBASE+"/stop", "application/json", c.Request.Body)

	c.Header("Access-Control-Allow-Origin", "*")
	c.Header("Access-Control-Allow-Headers", "access-control-allow-origin, access-control-allow-headers")
	if resp.StatusCode != 200 {
		c.JSON(resp.StatusCode, resp.Body)
		return
	}

	c.JSON(resp.StatusCode, Message{"Container stopped."})
}

func inspectContainer(c *gin.Context) {
	cID := c.Param("containerID")
	agentID := c.Param("agentID")
	a, err := mdb.GetAgent(agentID)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"Failed to load Agent:" + err.Error()})
		return
	}

	container := docker.Container{}
	resp, _ := http.Get("http://" + a.IP + ":" + a.Port + "/" + APIBASE + "/inspect/" + cID)

	decoder := json.NewDecoder(resp.Body)
	err = decoder.Decode(&container)
	if err != nil {
		c.JSON(http.StatusInternalServerError, ErrorResponse{"Could not retrieve container:" + err.Error()})
		return
	}

	c.Header("Access-Control-Allow-Origin", "*")
	c.JSON(http.StatusOK, container)
}

//index a humble welcome to a new user
func listContainers(c *gin.Context) {
	agents, err := mdb.Load()
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error while loading containers: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, agents)
}

func addContainers(c *gin.Context) {
	agent := Agent{}
	err := c.BindJSON(&agent)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error binding json: " + err.Error()})
		return
	}
	err = mdb.Save(agent)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error while saving container: " + err.Error()})
		return
	}

	c.JSON(http.StatusOK, Message{"Containers successfully saved."})

}
func deleteContainers(c *gin.Context) {
	agent := Agent{}
	err := c.BindJSON(&agent)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error binding json: " + err.Error()})
		return
	}
	err = mdb.Delete(agent)
	if err != nil {
		c.JSON(http.StatusBadRequest, ErrorResponse{"error while deleting containers: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, Message{"Containers successfully removed."})
}


agent/docker_commands.go
package main

import (
	"net/http"
	"strings"

	"github.com/fsouza/go-dockerclient"
	"github.com/gin-gonic/gin"
)

func stopContainer(c *gin.Context) {
	var container struct {
		ID string `json:"id"`
	}

	err := c.BindJSON(&container)
	if err != nil {
		e := ErrorResponse{}
		e.ErrorMessage = "error parsing json body:" + err.Error()
		c.JSON(http.StatusInternalServerError, e)
		return
	}
	//reflect.ValueOf(&t).MethodByName("Foo").Call([]reflect.Value{})
	endpoint := "unix:///var/run/docker.sock"
	client, _ := docker.NewClient(endpoint)

	err = client.StopContainer(container.ID, 10)
	if err != nil {
		e := ErrorResponse{}
		e.ErrorMessage = "error stopping containers:" + err.Error()
		c.JSON(http.StatusInternalServerError, e)
		return
	}
	m := Message{}
	m.Message = "continer stopped successfully"
	c.JSON(http.StatusOK, m)
}

func stopAllContainers(c *gin.Context) {

	var stopErrors [][]string
	for i := range stopErrors {
		stopErrors[i] = make([]string, 0)
	}
	//reflect.ValueOf(&t).MethodByName("Foo").Call([]reflect.Value{})
	endpoint := "unix:///var/run/docker.sock"
	client, _ := docker.NewClient(endpoint)
	runningContainers, err := client.ListContainers(docker.ListContainersOptions{All: false})

	if err != nil {
		e := ErrorResponse{}
		e.ErrorMessage = "error getting all containers:" + err.Error()
		c.JSON(http.StatusInternalServerError, e)
		return
	}

	for _, v := range runningContainers {
		err := client.StopContainer(v.ID, 1)
		if err != nil {
			stopErrors = append(stopErrors, v.Names)
		}
	}

	if len(stopErrors) != 0 {
		var errCon string
		for _, v := range stopErrors {
			errCon += "[" + strings.Join(v, "|") + "]"
		}
		e := ErrorResponse{}
		e.ErrorMessage = "error stopping containers:" + errCon
		c.JSON(http.StatusInternalServerError, e)
		return
	}

	m := Message{}
	m.Message = "all containers successfully stopped"
	c.JSON(http.StatusOK, m)
}

func inspectContainer(c *gin.Context) {
	cID := c.Param("id")
	endpoint := "unix:///var/run/docker.sock"
	client, _ := docker.NewClient(endpoint)
	con, err := client.InspectContainer(cID)
	if err != nil {
		e := ErrorResponse{"error while trying to inspect container:" + err.Error()}
		c.JSON(http.StatusInternalServerError, e)
	}

	c.JSON(http.StatusOK, con)
}


agent/agent.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"log"
	"net"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/fsouza/go-dockerclient"
	"github.com/gin-gonic/gin"
)

//APIBASE The base of the API that this agent provides
const APIBASE = "/api/1"

//Containers representing information about the running containers
type Containers struct {
	ID      string `json:"id"`
	Name    string `json:"name"`
	Command string `json:"command"`
	Port    string `json:"port"`
}

//Post represents a post to the server
type Post struct {
	AgentID             string       `json:"agentid"`
	ExpiredAfterSeconds int          `json:"expireAfterSeconds"`
	IPAddress           string       `json:"ip"`
	Port                string       `json:"port"`
	Containers          []Containers `json:"containers"`
}

var (
	serverURL          string
	agentID            string
	refreshRate        int
	expireAfterSeconds int
	port               string
)

func init() {
	flag.StringVar(&serverURL, "server", "http://localhost:8989", "The server uri where dockmaster is located.")
	flag.StringVar(&agentID, "agent", "localhost", "The name of an Agent. Example: TestQA1.")
	flag.IntVar(&refreshRate, "refresh", 60, "The rate at which this agent should check for changes in seconds.")
	flag.IntVar(&expireAfterSeconds, "expireAfterSeconds", 60, "The rate at which data sent by this agent should expire in seconds.")
	flag.StringVar(&port, "port", "9999", "The port number on which this agent is running on.")

	flag.Parse()
	go startDiscovering()
}

func main() {
	router := gin.Default()
	v1 := router.Group(APIBASE)
	{
		v1.POST("/stop", stopContainer)
		v1.POST("/stopAll", stopAllContainers)
		v1.GET("/inspect/:id", inspectContainer)
	}
	router.Run(":" + port)
}

func startDiscovering() {
	endpoint := "unix:///var/run/docker.sock"
	client, _ := docker.NewClient(endpoint)
	log.Println("Started listening... Refresh rate is :", refreshRate)
	for {
		log.Println("Assembling Post...")
		post := Post{AgentID: agentID, ExpiredAfterSeconds: expireAfterSeconds}
		containers := []Containers{}
		runningContainers, err := client.ListContainers(docker.ListContainersOptions{All: false})
		if err != nil {
			panic("Failed to connect to Docker Client." + err.Error())
		}
		for _, v := range runningContainers {
			c := Containers{}
			c.ID = v.ID
			c.Name = strings.Join(v.Names, ",")
			for _, p := range v.Ports {
				c.Port += p.IP + ":" + p.Type + ":" + strconv.Itoa(int(p.PrivatePort)) + ":" + strconv.Itoa(int(p.PublicPort))
			}
			c.Command = v.Command
			containers = append(containers, c)
		}
		post.Containers = containers
		post.IPAddress = getLocalIP()
		post.Port = port
		postString, err := json.Marshal(post)
		if err != nil {
			log.Println("Error occured while trying ot marshal POST:", err.Error())
		}
		req, err := http.NewRequest("POST", serverURL+"/api/1/add", bytes.NewBuffer(postString))
		if err != nil {
			log.Println("Failed to create post request... Trying again later.")
			time.Sleep(time.Second * time.Duration(refreshRate))
			continue
		}

		req.Header.Set("Content-Type", "application/json")
		log.Println("Posting JSON:", string(postString))
		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			log.Println("Failed to receive from server... Trying again later.")
			time.Sleep(time.Second * time.Duration(refreshRate))
			continue
		}
		defer resp.Body.Close()
		//TODO: Verify the response
		time.Sleep(time.Second * time.Duration(refreshRate))
	}
}

func getLocalIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		log.Println("Couldn't determine ip address. Sending empty string. Error:", err.Error())
		return ""
	}

	for _, a := range addrs {
		if ipnet, ok := a.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				return ipnet.IP.String()
			}
		}
	}

	return ""
}
